Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp

```diff
@@ -28,49 +28,355 @@
+#include "HTMLVideoElement.h"
+#include "Position.h"
 #include "SimpleRange.h"
+#include "TextBoundaries.h"
 #include "TextIterator.h"
 
 namespace WebCore {
 namespace FragmentDirectiveRangeFinder {
 
-static SimpleRange collapseIfRootsDiffer(SimpleRange&& range)
+enum class BoundaryPointIsAtEnd { No, Yes };
+enum class WordBounded : bool { No, Yes };
+
+static bool isSearchInvisible(const Node& node)
 {
-    return &range.start.container->rootNode() == &range.end.container->rootNode()
-        ? WTFMove(range) : SimpleRange { range.start, range.start };
+    // TODO: The computed value of its display property is none.
+    
+    // TODO: If the node serializes as void.
+    
+    if (is<HTMLIFrameElement>(node) || is<HTMLImageElement>(node) || is<HTMLMeterElement>(node) || is<HTMLObjectElement>(node) || is<HTMLProgressElement>(node) || is<HTMLStyleElement>(node) || is<HTMLScriptElement>(node) || is<HTMLVideoElement>(node) || is<HTMLAudioElement>(node))
+        return true;
+    
+    // TODO: Is a select element whose multiple content attribute is absent.
+    
+    return false;
 }
 
-static std::optional<SimpleRange> rangeOfString(const String& target, Document& document)
+static bool isNonSearchableSubtree(const Node& node)
 {
-    if (target.isEmpty())
+    const Node* traversingNode = &node;
+    while (traversingNode) {
+        if (isSearchInvisible(*traversingNode))
+            return true;
+        traversingNode = traversingNode->parentInComposedTree();
+    }
+    return false;
+}
+
+static const Node& nearestBlockAncestor(const Node& node)
+{
+    const Node* currentNode = &node;
+    while (currentNode) {
+        if (!currentNode->isTextNode() && currentNode->renderer() && currentNode->renderer()->style().isDisplayBlockLevel())
+            return *currentNode;
+        currentNode = currentNode->parentInComposedTree();
+    }
+    return node.document();
+}
+
+static std::optional<BoundaryPoint> getBoundaryPointAtIndex(unsigned index, Vector<Node*>& textNodeList, BoundaryPointIsAtEnd isEnd)
+{
+    unsigned currentLength = 0;
+    
+    for (auto* currentNode : textNodeList) {
+        unsigned nodeEnd = currentLength + currentNode->length();
+        if (isEnd == BoundaryPointIsAtEnd::Yes)
+            nodeEnd++;
+        if (nodeEnd > index)
+            return BoundaryPoint(*currentNode, index - currentLength);
+        currentLength += currentNode->length();
+    }
+    return std::nullopt;
+}
+
+static bool indexIsWordBoundary(const String& string, unsigned index)
+{
+    UBreakIterator* it = wordBreakIterator(string);
+    return ubrk_isBoundary(it, index);
+}
+
+static bool isVisibleTextNode(const Node& node)
+{
+    return node.isTextNode() && node.renderer() && node.renderer()->style().visibility() == Visibility::Visible;
+}
+
+static std::optional<SimpleRange> findRangeFromNodeList(const String& query, SimpleRange& searchRange, Vector<Node*>& nodes, WordBounded wordStartBounded, WordBounded wordEndBounded)
```

Comment:
Pleas don't use a vector of Node*. Use WeakPtr or Ref/RefPtr whichever is appropriate.